v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/source/reader/io/jdbc/rowmapper/JdbcSourceRowMapper.java

```diff
@@ -0,0 +1,100 @@
+import java.time.Instant;
+import java.util.concurrent.TimeUnit;
+import org.apache.avro.Schema;
+import org.apache.beam.sdk.io.jdbc.JdbcIO;
+import org.checkerframework.checker.initialization.qual.Initialized;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.checkerframework.checker.nullness.qual.UnknownKeyFor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Implement the {@link org.apache.beam.sdk.io.jdbc.JdbcIO.RowMapper} interface. */
+public final class JdbcSourceRowMapper implements JdbcIO.RowMapper<SourceRow> {
+
+  private final JdbcValueMappingsProvider mappingsProvider;
+  private final SourceTableSchema sourceTableSchema;
+
+  private static final Logger logger = LoggerFactory.getLogger(JdbcSourceRowMapper.class);
+
+  /**
+   * Construct {@link JdbcSourceRowMapper}.
+   *
+   * @param mappingsProvider Mapping Provider based on the type of database.
+   * @param sourceTableSchema Schema of source table.
+   */
+  public JdbcSourceRowMapper(
+      JdbcValueMappingsProvider mappingsProvider, SourceTableSchema sourceTableSchema) {
+    this.mappingsProvider = mappingsProvider;
+    this.sourceTableSchema = sourceTableSchema;
+  }
+
+  long getCurrentTime() {
+    Instant now = Instant.now();
+    long nanos = TimeUnit.SECONDS.toNanos(now.getEpochSecond()) + now.getNano();
+    return TimeUnit.NANOSECONDS.toMicros(nanos);
+  }
+
+  /**
+   * Map {@link java.sql.ResultSet} to {@link SourceRow}.
+   *
+   * @param resultSet the resultSet for a read record.
+   * @return SourceRow
+   * @throws Exception {@link ValueMappingException} when resultSet can't be extracted to expected
+   *     type. Typically, this indicates a change in source schema during migration.
+   */
+  @Override
+  public @UnknownKeyFor @Nullable @Initialized SourceRow mapRow(
+      @UnknownKeyFor @NonNull @Initialized ResultSet resultSet) throws Exception {
+    var builder = SourceRow.builder(sourceTableSchema, getCurrentTime());
+    this.sourceTableSchema
+        .sourceColumnNameToSourceColumnType()
+        .entrySet()
+        .forEach(
+            entry -> {
+              try {
+                Schema schema =
+                    this.sourceTableSchema.getAvroPayload().getField(entry.getKey()).schema();
+                // The Unified avro mapping produces a union of the mapped type with null type
+                // except for "Unsupported" case.
+                if (schema.isUnion()) {
+                  schema = schema.getTypes().get(1);
+                }
+                builder.setField(
+                    entry.getKey(),
+                    this.mappingsProvider
+                        .getMappings()
+                        .getOrDefault(entry.getValue().getName(), JdbcValueMapper.UNSUPPORTED)
+                        .mapValue(resultSet, entry.getKey(), schema));
+              } catch (SQLException e) {
+                logger.error(
```

Comment:
The error msg could be more informative. We can include details like the specific field and the expected data type to aid debugging.

Reply:
The resultSet exception will provide the necessary information.